### PR TITLE
Deploy volsync as mgdclusteraddon

### PR DESCRIPTION
**Pull request by tesshuflower**
will deploy volsync from the ACM hub by creating the appropriate volsync managedclusteraddon for the managed cluster that needs volsync.  will only deploy the managedclusteraddon if it doesn't already exist (allowing users to customize their own if they've already done this).

### CODE REVIEW

File: controllers/drplacementcontrolvolsync.go [hunk 1]
```diff
@@ -197,17 +201,9 @@ func (d *DRPCInstance) IsVolSyncReplicationRequired(homeCluster string) (bool, e
 		return false, fmt.Errorf("failed to find VRG on homeCluster %s", homeCluster)
 	}
 
-	if len(vrg.Status.ProtectedPVCs) == 0 {
```
Comment (BenamarMk) [inline controllers/drplacementcontrolvolsync.go:200]: We need to rework this function. As we discussed, we can't wait for `ProtectedPVCs` to be filled in, but at the same time, we need to return `WaitForSourceCluster` error on an empty `ProtectedPVCs`, otherwise, we may complete the action (deploy/failover/relocate) without setting up volsync. The next reconcile may set it up, but we can't depend on that as we don't know when it would be called.

One possible solution is to return an `WaitForSouceCluster` error [here](https://github.com/RamenDR/ramen/blob/01d4e722c466d6a6949f5dd021a20200c648f6d9/controllers/drplacementcontrol.go#L1396) instead. Something like:

```
        vrg := d.vrgs[homeCluster]
	if len(vrg.Status.ProtectedPVCs) == 0 {
		return WaitForSourceCluster
	}
        repReq := d.IsVolSyncReplicationRequired(homeCluster)
	...
```
In other words, just move that block from `IsVolSyncReplicationRequired` and put it in `EnsureCleanup`. Also the `IsVolSyncReplicationRequired` return value becomes a boolean return value only.

Comment (tesshuflower) [reply controllers/drplacementcontrolvolsync.go:200]: Thx Benamar, good catch - this should be addressed in the latest commit now: https://github.com/RamenDR/ramen/pull/529/commits/4229f1da9e644ace6039cfbf5f85bf9b14f8dac7

**Comment by ShyamsundarR** [comment]
If DRPC decides to install volsync on managed clusters that would then allow a user to create one just to install volsync on the clusters.

This should be installed using the DRPolicy/DRCluster resource instead, as these are cluster scoped and created by the cluster admin. So we should add a (say) bool to DRCluster to add volsync capability, and install the addon using the DRCluster reconciler. A DRPolicy will enable volsync if all clusters in its policy supports volsync.

**Comment by BenamarMk** [comment]
From gchat conversation:
---------------------------------------------
Shyam Ranganathan, 32 min
Ok here is what I see, PR https://github.com/RamenDR/ramen/pull/529 allows install of Volsync by a user, which I think should be based on DRCluster instead. If that is true, and DRCluster installs.configures the managed cluster anyway, we can add the volsync flag to the same to start using the Owns construct for the above problem

Benamar Mekhissi, 27 min
Are you suggesting for DRPC to add a flag to the DRCluster to install VolSync?

Meaning, when DRPC detects that VolSync is needed?

Shyam Ranganathan, 26 min
No, let the admin create DRCluster with the flag, only if that is true/available will DRPC use volsync for the specified policy

(again afk for 45 minutes... will be back around 4:15 PM)

Benamar Mekhissi, Now
Well, we talked about that. It is doable, but we wanted the user to be hands-free. IOW, if VolSync needs to be installed, then install it when needed.

**Comment by BenamarMk** [comment]
More conversation on gchat:
---------------------------------------
Shyam Ranganathan, 39 min
I think allowing user to install the operator would be a big problem (indirectly though). It is better to let the admin install/configure it for use

Benamar Mekhissi, 35 min
Yeah, we are all for that. Again, the goal was to install VolSync silently if and only if CephFS is the provisioner.

Shyam Ranganathan, 35 min
But we cannot merge DRPC deploying volsync, that may be a security/access issue

Benamar Mekhissi, 30 min
Is it really security issue? or just not nice to install something without Admin consent?

Shyam Ranganathan, 28 min
Definitely not nice to install without admin consent, the security or use aspect is present. If installed it opens up other namespaces to use Volsync (say natively etc.) which in turn may chew up bandwidth across clusters and not be desirable. It is something the admin should opt-in for IMO

I think we should do the following:
- merge https://github.com/RamenDR/ramen/pull/522 (5s reconcile till DataProtected)
- rework add-on from DRCluster/DRPolicy (or some such)
- rework the Owns to remove the 5s reconcile
Thoughts?

Benamar Mekhissi
Sounds good

Benamar Mekhissi, 26 min, Edited
Anyway, we can rework that. That shouldn't stop the QE to start testing if we have not closed on this by then.

Shyam Ranganathan, 25 min
I am also worried we are creating a storageClass with the clone PR, but we can use that as is for now, and make it a configurable entity (DRCluster/Policy) for CephFS like our replicationClass selector. I prefer not to add more RBAC as far as possible and also give a opt-out for the admin. As in this case as well we would create a SC when a user desires a DRPC for CephFS, than using one. (it is a pain to have too many pre-requisites as well, but some line in the sand...)

**Comment by ShyamsundarR** [comment]
Closing as this is addressed with https://github.com/RamenDR/ramen/pull/537
